src/xcoobee/sdk/Bees.js

```diff
@@ -1,247 +1,254 @@
+   * @async
+   * @param {Array<Object<string, Object>>} bees - A mapping of bee names to bee parameters.
+   * @param {string} bees[].key - The bee name.  A 'transfer' bee will be ignored.
+   * @param {Object} bees[].value - The bee parameters.
+   * @param {Object} options - The bee take off options.
+   * @param {Object} options.process -
+   * @param {Array<{value: any, name: string}>} [options.custom] - custom properties
+   * @param {Array<email|XcooBeeId>} [options.process.destinations] -
+   * @param {string[]} options.process.fileNames -
+   * @param {string} [options.process.userReference] -
+   * @param {TransactionSubscriptions} [subscriptions]
+   * @param {Config} [config] - If specified, the configuration to use instead of the
+   *   default.
+   *
+   * @returns {Promise<SuccessResponse | ErrorResponse>} - The response.
+   * @property {number} code - The response status code.
+   * @property {Error} [error] - The response error if status is not successful.
+   * @property {string} [error.message] - The error message.
+   * @property {string} request_id - The ID of the request generated by the XcooBee
+   *   system.
+   * @property {Object} [result] - The result of the response if status is successful.
+   * @property {string} [result.ref_id] - A reference ID generated by the XcooBee
+   *   system.
+   *
+   * @throws {XcooBeeError}
+   */
+  async takeOff(bees, options, subscriptions, config = null) {
+    this._assertValidState();
+    const apiCfg = SdkUtils.resolveApiCfg(config, this._.config);
+    const { apiKey, apiSecret, apiUrlRoot } = apiCfg;
+
+    const directiveInput = {
+      filenames: options.process.fileNames,
+      user_reference: options.process.userReference || null,
+    };
+
+    if (subscriptions) {
+      // TODO: validate subscriptions
+      directiveInput.subscriptions = subscriptions;
+    }
+
+    if (
+      options.custom
+      && Array.isArray(options.custom)
+      && options.custom.length
+    ) {
+      directiveInput.custom = options.custom.slice();
+    }
+
+    if (
+      Array.isArray(options.process.destinations)
+      && options.process.destinations.length > 0
+    ) {
+      directiveInput.destinations = options.process.destinations.map(
+        (destination) => {
+          if (ApiUtils.appearsToBeAnEmailAddress(destination)) {
+            return { email: destination };
+          }
+          return { xcoobee_id: destination };
+        }
+      );
+    }
+
+    directiveInput.bees = Object
+      .keys(bees)
+      .map((beeName) => {
+        if (beeName !== 'transfer') {
+          return { bee_name: beeName, params: JSON.stringify(bees[beeName]) };
+        }
+        return null;
+      })
+      .filter(bee => bee);
+
+    try {
+      const apiAccessToken = await this._.apiAccessTokenCache.get(apiUrlRoot, apiKey, apiSecret);
+      const ref_id = await DirectiveApi.addDirective(apiUrlRoot, apiAccessToken, directiveInput);
+      return new SuccessResponse({ ref_id });
```

Comment:
Note: The reason for assigning a result to a variable that is immediately returned is to make debugging easier.

With

```
  const result = new SuccessResponse(...);
  return result;
```
I can set a breakpoint on the return statement and have easy access to the result.

With
```
  return new SuccessResponse(...);
```
I have to set a breakpoint here, step into the constructor, step out of the constructor and into the code that called this function. Then I can examine the local variables.

Much easier to debug with two lines instead of one.

Please let me know if you know of a way to quickly see the result without multiple steps.

Reply:
you won`t be able to debug it in production anyway, as code should be minified
will revert it, but I think it is unnecessary slow down

Reply:
The SDK is designed to run in both Node and the Browser. So, it won't necessarily be minified for Node. We do minify the browser version but I believe we also generate source maps too. I haven't tried to debug in production with source maps. Do you know if you can still set breakpoints on specific lines even though the source is minified to just one line? I am mainly concerned about debugging during development, not production.

As I learned throughout the years, I have found that developer time and experience almost always outweighs very minor performance improvements. Also, in a project like this one where we have network calls that we wait on before performing more processing, minor improvements like 1 less local variable are negligible. One network call is likely thousands to millions of times slower than the time difference of any performance gain by eliminating one local variable.

One of the principles of software development is to never pre-optimize. And I think it is best to couple any optimization with performance profiling where you take measurements before and after so know how much of an improvement (and in some cases may find out what you believed to be an improvement turns out to slow things down). The profiling can also help identify the biggest areas of concern which you should focus on first.

With that said, I am not saying I don't sometimes do minor performance optimizations without a performance analysis first. There are some code styles that I know have a little bit better performance and I'll blindly do them as long as it won't have a negative affect on developer experience (DX).

Reply:
Team:
We should attempt to make debugging easier. 
Our goal should be to code for humans first.